File: FinacPlus/coding task challenge/coding_challenge_Python/optimized_units.py
def optimized_units(units,max_value):
    total_units_used = 0
    unit_counts = [0] * (max_value + 1)

    # Dynamic programming approach to find the minimum number of units for each value
    for value in range(1, max_value + 1):
        min_units = float('inf')
        for unit in units:
            if value >= unit:
                min_units = min(min_units, unit_counts[value - unit] + 1)
        unit_counts[value] = min_units
        total_units_used += min_units

    # Calculate average units used
    average_units = total_units_used / max_value

    # Print results
    print(f"Units used: {units}")
    print(f"Average units used: {average_units}")

print("Enter the values: ")
units = list(map(int,input().split()))
max_value = int(input())
# Run the function
optimized_units(units,max_value)

File: FinacPlus/coding task challenge/coding_challenge _Java/OptimizedUnits.java
import java.util.*;

public class OptimizedUnits {
    
    public static void main(String[] args) {
        Scanner sc =  new Scanner(System.in);
        System.out.print("Enter the size of Units : ");
        int n = sc.nextInt();
        int[] units = new int[n];
        System.out.print("Enter the values of Units : ");
        for(int i=0;i<n;i++)
            units[i] = sc.nextInt();
        System.out.print("Enter the max_value : ");
        int maxValue = sc.nextInt();
        int totalUnitsUsed = 0;
        int[] unitCounts = new int[maxValue + 1];

        // Dynamic programming approach to find the minimum number of units for each value
        for (int value = 1; value <= maxValue; value++) {
            int minUnits = Integer.MAX_VALUE;
            for (int unit : units) {
                if (value >= unit) {
                    minUnits = Math.min(minUnits, unitCounts[value - unit] + 1);
                }
            }
            unitCounts[value] = minUnits;
            totalUnitsUsed += minUnits;
        }

        // Calculate average units used
        double averageUnits = (double) totalUnitsUsed / maxValue;

        // Print results
        System.out.println("Units used: " + Arrays.toString(units));
        System.out.println("Average units used: " + averageUnits);
    }
}